docs/proposals/decouple-application-sync-user-using-impersonation

```diff
@@ -0,0 +1,321 @@
+- Fix Application Controller sync.go to set the Impersonate configuration from the Application CR to the SyncContext Object (rawConfig and restConfig field, need to understand which config is used for the actual sync and if both configs need to be impersonated.)
+#### Component: ArgoCD UI
+
+- Provide option to provide `serviceAccountName` for a `Destination` section in an `AppProject` created/updated via the web console.
+- Update the User Guide documentation on how to use these newly added fields from the web console.
+#### Component: ArgoCD CLI
+
+- Provide option to provide serviceAccountName for an Application created/updated via the console.
+- Provide option to provide defaultServiceAccount for an AppProject created/updated via the web console.
+- Update the User Guide and other documentation where the CLI option usages are explained.
+
+
+### Detailed examples
+
+#### Example 1: Service account for application sync specified at the AppProject level for all namespaces
+In this specific scenario, service account name `generic-deployer` will get used for the application sync as the namespace `guestbook` matches the regex `*`.
+
+```
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: guestbook
+  namespace: argocd
+spec:
+  project: my-project
+  source:
+    repoURL: https://github.com/argoproj/argocd-example-apps.git
+    targetRevision: HEAD
+    path: guestbook
+  destination:
+    server: https://kubernetes.default.svc
+    namespace: guestbook
+---
+apiVersion: argoproj.io/v1alpha1
+kind: AppProject
+metadata:
+  name: my-project
+  namespace: argocd
+  finalizers:
+    - resources-finalizer.argocd.argoproj.io
+spec:
+  description: Example Project
+  # Allow manifests to deploy from any Git repos
+  sourceRepos:
+  - '*'
+  destinations:
+  - namespace: *
+    server: https://kubernetes.default.svc
+    serviceAccountName: generic-deployer
+```
+
+#### Example 2: Service account for application sync specified at the AppProject level for specific namespaces
+
+In this specific scenario, service account name `guestbook-deployer` will get used as it matches to the specific namespace `guestbook`.
+```
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: guestbook
+  namespace: argocd
+spec:
+  project: my-project
+  source:
+    repoURL: https://github.com/argoproj/argocd-example-apps.git
+    targetRevision: HEAD
+    path: guestbook
+  destination:
+    server: https://kubernetes.default.svc
+    namespace: guestbook
+---
+apiVersion: argoproj.io/v1alpha1
+kind: AppProject
+metadata:
+  name: my-project
+  namespace: argocd
+  finalizers:
+    - resources-finalizer.argocd.argoproj.io
+spec:
+  description: Example Project
+  # Allow manifests to deploy from any Git repos
+  sourceRepos:
+  - '*'
+  destinations:
+  - namespace: *
+    server: https://kubernetes.default.svc
+    serviceAccountName: generic-deployer
```

Comment:
One question that just came to me is what happens if there are multiple potential matches, i.e there are several potential serviceaccounts that gets returned? Does the first match win, or should this be an error?

Reply:
How should negated destinations be ranked? For example if we have a destination looking like

```
  destinations:
  - namespace: !something*
    server: !https://kubernetes.default.svc
    serviceAccountName: generic-deployer
  - namespace: not-something
    server: https://server-1
    serviceAccountName: some-other-deployer
```

How should that be ranked?

Reply:
I am not sure how we should handle negative matches. Can that be prioritized the least ? or is it better to keep it simple and throw an error if there are multiple matches ?

Reply:
IMO I think it's better to keep it simple and throw an error on multiple matches, WDYT @crenshaw-dev?

Reply:
Keeping the logic to throw error if there are multiple matches either through regex or an exact match of the destination.

Reply:
I think it might actually be kinda weird to have the `destinations` field serve two purposes: allowing/denying clusters/namespaces and also associating service account names with clusters.

What if we just add a new field called clusterServiceAccounts?

```yaml
clusterServiceAccounts:
- server: https://kubernetes.default.svc # ! negation is not allowed here
  serviceAccountName: in-cluster
- server: * # Fallback, first match wins
  serviceAccountName: argocd-default-sa
```

I don't wanna overload the `destinations` field with too much functionality. Causes confusion.